### PR TITLE
[SPARK-24107][CORE][followup] ChunkedByteBuffer.writeFully method has not reset the limit value

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

According to the discussion in https://github.com/apache/spark/pull/21175 , this PR proposes 2 improvements:
1. add comments to explain why we call `limit` to write out `ByteBuffer` with slices.
2. remove the `try ... finally`

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @manbuyun @JoshRosen @jiangxb1987

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 1]
```diff
@@ -63,15 +63,18 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
...
+      val originalLimit = bytes.limit()
       while (bytes.hasRemaining) {
-        try {
-          val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
-          bytes.limit(bytes.position() + ioSize)
-          channel.write(bytes)
-        } finally {
-          bytes.limit(curChunkLimit)
-        }
+        // If `bytes` is an on-heap ByteBuffer, the JDK will copy it to a temporary direct
```
Comment (advancedxy) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: how about `the JDK` -> `the JVM`?

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 2]
```diff
@@ -63,15 +63,18 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
...
+        // ByteBuffer when writing it out. The JDK caches one temporary buffer per thread, and we
```
Comment (advancedxy) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: > The JDK caches one temporary buffer per thread

I don't think this statement is correct.  According to [Util.java](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/nio/ch/Util.java#l48), the cached  number of temporary direct buffer is up to `IOUtil.IOV_MAX`. 

> if the cached temp buffer gets created and freed frequently

The problem is that the varied-sized heap buffers could cause a new allocation of temporary direct buffer and free of old direct buffer if the buffer size is larger than before

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala (hunk 1 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: the caching happens in the JDK code, not some magic inside JVM.

**Comment by JoshRosen** [comment]
LGTM. Thanks for these changes; they really help to clarify this tricky piece of code for readers.

**Comment by cloud-fan** [comment]
retest this please

**Comment by viirya** [comment]
LGTM. Good comment added.

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**